### PR TITLE
Addition of a shell script to automate CI workflow

### PR DESCRIPTION
**Pull request by AbhijithGanesh**
## Description

Fixes pyrsia/pyrsia#437

This PR will simplify the pre-commit process, the developer can run locally. The pre-commit script to check if their code base conforms to the requirements of the project.

It checks the essential features such as:
- cargo audit
- cargo clippy
- cargo test
- rustup fmt 

The `pre-commit.sh` is for Linux/UNIX and `pre-commit.bat` is for Windows.

Signed-off-by: Abhijith Ganesh <67182544+AbhijithGanesh@users.noreply.github.com>

## PR Checklist

- [x] I've built the code `cargo build`. For major changes, `cargo build --workspace --release` is recommended.
- [x] I've run the automated unit tests `cargo test`. This executes our automated unit tests.
- [x] I've not broken any existing tests or functionality. In addition to `cargo test`, I've run `cargo clippy`
- [x] I've not introduced any new known security vulnerabilities. I've run `cargo audit`

- [x] I've included a brief description and a good title of the proposed changes and how to test them.
- [x] I've read the [contributing guidelines](https://github.com/pyrsia/.github/blob/main/contributing.md).
- [x] I've associated an [issue](https://github.com/pyrsia/pyrsia/issues) with this Pull Request. 
- [x] I've checked that the code is up-to-date with the `pyrsia/main` branch.
- [x] I've read ["What is a Good PR?"](https://github.com/pyrsia/pyrsia/blob/main/docs/good_pr.md)
- [x] I've assigned this Pull Request to "pyrsia/collaborators"

### CODE REVIEW

File: pre-commit.bat [hunk 1]
```diff
@@ -0,0 +1,6 @@
+cargo install cargo-audit;
+cargo audit;
+cargo clippy;   
+rustup component add rustfmt;
+cargo fmt --check;
+cargo "test";
```
Comment (prince-chrismc) [inline pre-commit.bat:6]:
```suggestion
cargo test --workspace;

```

File: pre-commit.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# !/bin/sh
+cargo install cargo-audit;
+cargo audit;
+cargo clippy;   
+rustup component add rustfmt;
+cargo fmt --check;
+cargo "test";
```
Comment (prince-chrismc) [inline pre-commit.sh:7]:
```suggestion
cargo test --workspace;

```

**Review by prince-chrismc — COMMENTED** [review]
Some nit picking

**Comment by prince-chrismc** [comment]
🤔 Seems to be missing the "build" step, can you point it out incase I missed it?
Or please add it!

**Comment by prince-chrismc** [comment]
> I thought build was considered out of the scope of the script because that is whole a lot diff

The goal is to have a script that "recreates" the primary CI pipeline to help bootstrap new contributors

File: pre-commit.sh [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# !/bin/sh
+echo "Welcome to the Pyrsia local tests.\nThis might take sometime, donot interrupt it if the screen is blank."
+cargo install cargo-audit;
+cargo audit;
+cargo clippy;   
+rustup component add rustfmt;
+cargo fmt --check;
+cargo "test --workspace";
```
Comment (prince-chrismc) [inline pre-commit.sh:8]: extra quotes

```suggestion
cargo test --workspace;
```

**Review by prince-chrismc — COMMENTED** [review]
Two small nits

- qutoes
- end of file line 
https://github.com/pyrsia/pyrsia/blob/48af701ab14e3b32ba44caf47968a45684737506/.editorconfig#L6

And please add the `cargo build --workspace` we really want that

File: pre-commit.sh (hunk 3 above)
Comment (AbhijithGanesh) [reply pre-commit.sh:8]: I had to make those quotes because the word test is a command in some shells

**Review by prince-chrismc — APPROVED** [review]
LGTM 🚀

File: pre-commit.bat [hunk 4]
```diff
@@ -0,0 +1,9 @@
+echo "Welcome to the Pyrsia local tests.\nThis might take sometime, donot interrupt it if the screen is blank."
```
Comment (efrisch) [inline pre-commit.bat:1]: `@echo off`

File: pre-commit.sh [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# !/bin/sh
```
Comment (efrisch) [inline pre-commit.sh:1]: `#!/usr/bin/env bash`

File: pre-commit.bat [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+cargo clean;
```
Comment (efrisch) [inline pre-commit.bat:2]: You don't need the semicolons at the end of line.

File: pre-commit.sh [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+echo "Welcome to the Pyrsia local tests.\nThis might take sometime, donot interrupt it if the screen is blank."
+cargo clean;
```
Comment (efrisch) [inline pre-commit.sh:3]: Don't need the semicolons at the end of line.

File: pre-commit.sh [hunk 8]
```diff
@@ -0,0 +1,10 @@
+# !/bin/sh
+echo "Welcome to the Pyrsia local tests.\nThis might take sometime, donot interrupt it if the screen is blank."
```
Comment (efrisch) [inline pre-commit.sh:2]: In general, I would prefer `printf` over `echo`. You also need to make sure you're using `bash` for that to be reliable. https://www.makeuseof.com/bash-printf-examples/

**Review by efrisch — APPROVED** [review]
👍

File: pre-commit.sh (hunk 5 above)
Comment (AbhijithGanesh) [reply pre-commit.sh:1]: Shell is more universal

File: pre-commit.bat (hunk 4 above)
Comment (AbhijithGanesh) [reply pre-commit.bat:1]: Why though?

Comment (efrisch) [reply pre-commit.bat:1]: As the first line it prevents the commands in the batch file from being echo'd before execution.

File: pre-commit.sh (hunk 5 above)
Comment (efrisch) [reply pre-commit.sh:1]: On ubuntu `/bin/sh` is dash; not bash. And the Debian Almquist Shell is not more universal. On mac `/bin/sh` is bash. https://wiki.ubuntu.com/DashAsBinSh

**Pull request merged**